src/cached_object_store.rs

```diff
@@ -0,0 +1,1074 @@
+        let part_path = Self::make_part_path(
+            self.root_folder.clone(),
+            &self.location,
+            part_number,
+            self.part_size,
+        );
+
+        // if the part file does not exist, return None
+        let exists = fs::try_exists(&part_path).await.unwrap_or(false);
+        if !exists {
+            return Ok(None);
+        }
+
+        let file = File::open(&part_path).await.map_err(wrap_io_err)?;
+        let mut reader = tokio::io::BufReader::new(file);
+        let mut buffer = Vec::new();
+        reader.read_to_end(&mut buffer).await.map_err(wrap_io_err)?;
+        Ok(Some(Bytes::from(buffer)))
+    }
+
+    #[cfg(test)]
+    async fn cached_parts(&self) -> object_store::Result<Vec<PartID>> {
+        let head_path = Self::make_head_path(self.root_folder.clone(), &self.location);
+        let directory_path = match head_path.parent() {
+            Some(directory_path) => directory_path,
+            None => return Ok(vec![]),
+        };
+        let target_prefix = "_part";
+
+        let mut entries = match fs::read_dir(directory_path).await {
+            Ok(entries) => entries,
+            Err(err) => {
+                if err.kind() == std::io::ErrorKind::NotFound {
+                    return Ok(vec![]);
+                } else {
+                    return Err(wrap_io_err(err));
+                }
+            }
+        };
+
+        let mut part_file_names = vec![];
+        while let Some(entry) = entries.next_entry().await.map_err(wrap_io_err)? {
+            let metadata = entry.metadata().await.map_err(wrap_io_err)?;
+            if metadata.is_dir() {
+                continue;
+            }
+            let file_name = entry.file_name();
+            let file_name_str = file_name.to_string_lossy();
+            if file_name_str.starts_with(target_prefix) {
+                part_file_names.push(file_name_str.to_string());
+            }
+        }
+
+        // not cached at all
+        if part_file_names.is_empty() {
+            return Ok(vec![]);
+        }
+
+        // sort the paths in alphabetical order
+        part_file_names.sort();
+
+        // retrieve the part numbers from the paths
+        let mut part_numbers = Vec::with_capacity(part_file_names.len());
+        for part_file_name in part_file_names.iter() {
+            let part_number = part_file_name
+                .split('-')
+                .last()
+                .and_then(|part_number| part_number.parse::<usize>().ok());
+            if let Some(part_number) = part_number {
+                part_numbers.push(part_number);
+            }
+        }
+
+        Ok(part_numbers)
+    }
+
+    async fn save_head(&self, head: (&ObjectMeta, &Attributes)) -> object_store::Result<()> {
+        // if the meta file exists and not corrupted, do nothing
+        match self.read_head().await {
+            Ok(Some(_)) => return Ok(()),
```

Comment:
Why doesn't this do fs::exists, like save_part? Why does it read_head?

Reply:
besides checking existence, `read_head()` also help checks whether the file is corrupted or not.

i guess this can be replaced with checksums in a unified manner.